core/envelope-sim/src/main/java/fr/sncf/osrd/envelope/part/EnvelopePart.kt

```diff
@@ -0,0 +1,677 @@
+        // the index of the step is the index of the point which starts the range
+        val stepIndex = insertionPoint - 1
+        return EnvelopePhysics.interpolateStepSpeed(
+            positions[stepIndex],
+            positions[stepIndex + 1],
+            speeds[stepIndex],
+            speeds[stepIndex + 1],
+            position - positions[stepIndex]
+        )
+    }
+
+    /** Given a position return the interpolated speed. */
+    fun interpolateSpeed(stepIndex: Int, position: Double): Double {
+        assert(checkPosition(stepIndex, position))
+        if (position == positions[stepIndex]) return speeds[stepIndex]
+        if (position == positions[stepIndex + 1]) return speeds[stepIndex + 1]
+        return EnvelopePhysics.interpolateStepSpeed(
+            positions[stepIndex],
+            positions[stepIndex + 1],
+            speeds[stepIndex],
+            speeds[stepIndex + 1],
+            position - positions[stepIndex]
+        )
+    }
+
+    /** Given a position return the interpolated deltaTime */
+    private fun interpolateTimeDelta(stepIndex: Int, position: Double): Double {
+        assert(checkPosition(stepIndex, position))
+        if (position == positions[stepIndex]) return 0.0
+        if (position == positions[stepIndex + 1]) return timeDeltas[stepIndex]
+        return EnvelopePhysics.interpolateStepTime(
+            positions[stepIndex],
+            positions[stepIndex + 1],
+            speeds[stepIndex],
+            speeds[stepIndex + 1],
+            position - positions[stepIndex]
+        )
+    }
+
+    /**
+     * Returns the time required to get from the start of the envelope part to the given position,
+     * in microseconds.
+     */
+    fun interpolateTotalTimeUS(position: Double): Long {
+        assert(position >= beginPos)
+        assert(position <= endPos)
+        val pointIndex = Arrays.binarySearch(positions, position)
+        // if the position matches one of the data points
+        if (pointIndex >= 0) return getTotalTimeUS(pointIndex)
+
+        // when the position isn't found, binarySearch returns -(insertion point) - 1
+        val insertionPoint = -(pointIndex + 1)
+        // the index of the step is the index of the point which starts the range
+        val stepIndex = insertionPoint - 1
+        val timeToStepStart = getTotalTimeUS(stepIndex)
+        val interpolatedTime = interpolateTimeDelta(stepIndex, position)
+        return timeToStepStart + (interpolatedTime * 1000000).toLong()
+    }
+
+    /**
+     * Returns the time required to get from the start of the envelope part to the given position,
+     * in seconds
+     */
+    fun interpolateTotalTime(position: Double): Double {
+        return (interpolateTotalTimeUS(position).toDouble()) / 1000000
+    }
+
+    /**
+     * Given a speed return a position. The envelopePart must be bijective in order for this method
+     * to work
+     */
+    fun interpolatePosition(speed: Double): Double? {
```

Comment:
It's weird that we return a nullable, considering we assert that the inputs would give a valid solution